connector/src/main/java/org/jboss/as/connector/subsystems/jca/JcaWorkManagerDefinition.java

```diff
@@ -71,11 +84,57 @@ public void registerAttributes(ManagementResourceRegistration resourceRegistrati
 
     @Override
         public void registerChildren(ManagementResourceRegistration resourceRegistration) {
-        resourceRegistration.registerSubModel(BoundedQueueThreadPoolResourceDefinition.create(WORKMANAGER_SHORT_RUNNING, ThreadsServices.STANDARD_THREAD_FACTORY_RESOLVER, ThreadsServices.STANDARD_HANDOFF_EXECUTOR_RESOLVER,
-                ThreadsServices.EXECUTOR.append(WORKMANAGER_SHORT_RUNNING), registerRuntimeOnly));
-        resourceRegistration.registerSubModel(BoundedQueueThreadPoolResourceDefinition.create(WORKMANAGER_LONG_RUNNING, ThreadsServices.STANDARD_THREAD_FACTORY_RESOLVER, ThreadsServices.STANDARD_HANDOFF_EXECUTOR_RESOLVER,
-                        ThreadsServices.EXECUTOR.append(WORKMANAGER_LONG_RUNNING), registerRuntimeOnly));
+        registerSubModels(resourceRegistration, registerRuntimeOnly);
+    }
+
+    static void registerSubModels(ManagementResourceRegistration resourceRegistration, boolean runtimeOnly) {
+        final BoundedQueueThreadPoolAdd shortRunningThreadPoolAdd = new BoundedQueueThreadPoolAdd(false,
+                ThreadsServices.STANDARD_THREAD_FACTORY_RESOLVER, ThreadsServices.STANDARD_HANDOFF_EXECUTOR_RESOLVER,
+                ThreadsServices.EXECUTOR.append(WORKMANAGER_SHORT_RUNNING)) {
+            @Override
+            protected void populateModel(final OperationContext context, final ModelNode operation, final Resource resource)
+                    throws OperationFailedException {
+                checkThreadPool(context, operation, WORKMANAGER_SHORT_RUNNING);
+                super.populateModel(context, operation, resource);
+            }
+        };
+        resourceRegistration.registerSubModel(
+                new JCAThreadPoolResourceDefinition(false, runtimeOnly, WORKMANAGER_SHORT_RUNNING, ThreadsServices.EXECUTOR.append(WORKMANAGER_SHORT_RUNNING),
+                        CommonAttributes.BOUNDED_QUEUE_THREAD_POOL, shortRunningThreadPoolAdd, new BoundedQueueThreadPoolRemove(shortRunningThreadPoolAdd)));
+
+        final BoundedQueueThreadPoolAdd longRunningThreadPoolAdd = new BoundedQueueThreadPoolAdd(false,
+                ThreadsServices.STANDARD_THREAD_FACTORY_RESOLVER, ThreadsServices.STANDARD_HANDOFF_EXECUTOR_RESOLVER,
+                ThreadsServices.EXECUTOR.append(WORKMANAGER_LONG_RUNNING)) {
+            @Override
+            protected void populateModel(final OperationContext context, final ModelNode operation, final Resource resource)
```

Comment:
I think I would do this in an additional step, pseudo-code-ish:
```
context.addStep(new OperationsStepHandler(){
    public void execute(OC oc, ModelNode op) {
        checkThreadPool();
   }
}, MODEL);
```

This extra step will then happen at the end of the MODEL stage, and guards against a composite operation doing something weird like adding and removing this a lot of times when the end result is that there is only one (which should work).